qiita_db/util.py

```diff
@@ -2311,3 +2317,337 @@ def send_email(to, subject, body):
+
+    ax.set_xscale('log')
+    ax.set_yscale('log')
+    ax.set_ylabel(curr)
+    ax.set_xlabel("samples * columns")
+
+    best_model, options = _resource_allocation_calculate(_df, x_data, y_data,
+                                                         models, curr, 100)
+    k, a, b = options.x
+    x_plot = np.array(sorted(_df[COL_NAME].unique()))
+
+    if curr == "MaxRSSRaw":
+        y_plot = M1G*1.5 + (np.log(x_plot)*M1G*1.5) + (x_plot*3500)
+        ax.plot(x_plot, y_plot, linewidth=1, color='red',
+                label="predefined curve")
+    if curr == "ElapsedRaw":
+        y_plot = 600 + (np.log(x_plot)*3000)
+        ax.plot(x_plot, y_plot, linewidth=1, color='red',
+                label="predefined curve")
+
+    y_plot = best_model(x_plot, k, a, b)
+    ax.plot(x_plot, y_plot, linewidth=1, color='orange')
+
+    maxi = naturalsize(_df[curr].max(), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(_df[curr].max()))
+    cmax = naturalsize(max(y_plot), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(max(y_plot)))
+
+    mini = naturalsize(_df[curr].min(), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(_df[curr].min()))
+    cmin = naturalsize(min(y_plot), gnu=True) if curr == "MaxRSSRaw" else \
+        timedelta(seconds=float(min(y_plot)))
+
+    x_plot = np.array(_df[COL_NAME])
+    failures_df = _resource_allocation_failures(_df, k, a, b, best_model, curr)
+    failures = failures_df.shape[0]
+
+    ax.scatter(failures_df[COL_NAME], failures_df[curr], color='red', s=3,
+               label="failures")
+
+    ax.set_title(f'{cname}: {sname}\n real: {mini} || {maxi}\n'
+                 f'calculated: {cmin} || {cmax}\n'
+                 f'failures: {failures}')
+    ax.legend(loc='upper left')
+    return best_model, options
+
+
+def _resource_allocation_calculate(_df, x, y, models, type_, depth):
+    """Helper function for resource allocation plot. Calculates best_model and
+    best_result given the models list and x,y data.
+
+    Parameters
+    ----------
+    x: pandas.Series (pandas column), required
+        Represents x data for the function calculation
+    y: pandas.Series (pandas column), required
+        Represents y data for the function calculation
+    models: list, required
+        List of functions that will be used for visualization
+    depth: int, required
+        Maximum number of iterations in binary search
+
+    Returns
+    ----------
+    best_model: function
+        best fitting function for the current list models
+    best_result: object
+        object containing constants for the best model (e.g. k, a, b in kx+b*a)
+    """
+
+    init = [1, 1, 1]
+    best_model = None
+    best_result = None
+    best_failures = np.inf
+    best_max = np.inf
+    counter = 0
+    for model in models:
+        counter += 1
+        sl = 2000
+        sr = 100000
+        left = sl
+        right = sr
```

Comment:
Can you add a small comment explaining what are these?

Reply:
These are initialization constants for binary search, where left is 2000 and right is 100000 at the start. I define them separately from left and right because I run the search on different models. Binary Search tries to find smallest penalty weight for which the curve will have least errors.